template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateProcessor.java

```diff
@@ -291,6 +311,52 @@ public String getStackVersion() {
         return cmTemplate.getCdhVersion();
     }
 
+    @Override
+    public Map<String, Map<String, ServiceAttributes>> getHostGroupBasedServiceAttributes() {
+        // The only attributes required at the moment relate to the YARN NodeManager.
+        // This functionality can be moved to an interface / separated if additional components
+        // need this functionality.
+        Map<String, Set<ServiceComponent>> hgToNonGwServiceComponents = getNonGatewayServicesByHostGroup();
+        Map<String, Set<ServiceComponent>> hgToNonGwServiceComponentsWithYarnNMs = new HashMap<>();
+        for (Entry<String, Set<ServiceComponent>> entry : hgToNonGwServiceComponents.entrySet()) {
+            for (ServiceComponent sc : entry.getValue()) {
+                if (sc.getService().equalsIgnoreCase(YarnRoles.YARN) && sc.getComponent()
+                        .equalsIgnoreCase(YarnRoles.NODEMANAGER)) {
+                    hgToNonGwServiceComponentsWithYarnNMs.put(entry.getKey(), entry.getValue());
+                    break;
+                }
+            }
+        }
+        // We have the hostGroups without GATEWAY components with YARN NodeManagers
+        Map<String, Set<String>> componentsByHostGroup = hgToNonGwServiceComponentsWithYarnNMs.entrySet()
+                .stream().collect(
+                        toMap(
+                                Entry::getKey,
+                                e -> e.getValue().stream()
+                                .map(ServiceComponent::getComponent)
+                                .collect(Collectors.toUnmodifiableSet())
+                ));
+        // Re-using the current LoadBasedAutoScaling recommendation along with the above YARN filter.
+        Set<String> computeHostGroups = getRecommendationByBlacklist(BlackListedLoadBasedAutoscaleRole.class,
+                true, componentsByHostGroup);
+
+        Map<String, Map<String, ServiceAttributes>> result = new HashMap<>();
+
+        for (String hg : componentsByHostGroup.keySet()) {
+            Map<String, ServiceAttributes> m;
+            if (computeHostGroups.contains(hg)) {
+                m = Collections.singletonMap(YarnRoles.YARN, new ServiceAttributes(ServiceComponent.of(YarnRoles.YARN, YarnRoles.NODEMANAGER),
+                        Collections.singletonMap(YarnConstants.ATTRIBUTE_NAME_NODE_INSTANCE_TYPE, YarnConstants.ATTRIBUTE_NODE_INSTANCE_TYPE_COMPUTE)));
+            } else {
+                m = Collections.singletonMap(YarnRoles.YARN, new ServiceAttributes(ServiceComponent.of(YarnRoles.YARN, YarnRoles.NODEMANAGER),
+                        Collections.singletonMap(YarnConstants.ATTRIBUTE_NAME_NODE_INSTANCE_TYPE, YarnConstants.ATTRIBUTE_NODE_INSTANCE_TYPE_WORKER)));
+            }
+            result.put(hg, m);
+        }
+        LOGGER.debug("ServiceAttributes: {}", result);
```

Comment:
```        Map<String, Map<String, ServiceAttributes>> result = new HashMap<>();
        for (String hg : componentsByHostGroup.keySet()) {
            String instanceType = computeHostGroups.contains(hg) ? YarnConstants.ATTRIBUTE_NODE_INSTANCE_TYPE_COMPUTE
                    : YarnConstants.ATTRIBUTE_NODE_INSTANCE_TYPE_WORKER;
            Map<String, ServiceAttributes> m = Collections.singletonMap(YarnRoles.YARN, new ServiceAttributes(ServiceComponent.of(YarnRoles.YARN, YarnRoles.NODEMANAGER),
                    Collections.singletonMap(YarnConstants.ATTRIBUTE_NAME_NODE_INSTANCE_TYPE, instanceType)));
            result.put(hg, m);
        }
```
seems a bit readable this way and could be moved to a separate method also

Reply:
Also `m` is not a good variable name, please rename it to something more helpful